pkg/storegateway/bucket_index_reader.go

```diff
@@ -413,9 +413,9 @@ func (r *bucketIndexReader) fetchPostings(ctx context.Context, keys []labels.Lab
 				// Return postings and fill LRU cache.
 				// Truncate first 4 bytes which are length of posting.
 				output[p.keyID] = newBigEndianPostings(pBytes[4:])
```

Comment:
Why do we need a lock around this access too? It is a slice and I assume we only assign each element once

Reply:
Follow up PR :)

Reply:
The answer is that we need to protect `output` from concurrency, so the lock is required. But we don't need a per `bucketIndexReader` mutex: we can just have a local one. In the next PR I will get rid of `bucketIndexReader.mtx` completely.

Reply:
You're right. Unless bugs in the postings resizing, there's no concurrent access for a given key and the slices is preallocated. I'm convinced we can remove it, but I still would prefer to do it in a separate PR (the one removing `mtx` at all) given this PR focus is "Do not call caching functions with mutex in bucketIndexReader".

Reply:
See: https://github.com/grafana/mimir/pull/3500